### PR TITLE
[KEW-43] proof of readiness scheduling

### PR DESCRIPTION
**Pull request by jalbiero**

### CODE REVIEW

File: contracts/eosio.system/src/producer_pay.cpp [hunk 1]
```diff
@@ -43,7 +43,7 @@ namespace eosiosystem {
 
       /// only update block producers once every minute, block_timestamp is in half seconds
       if( timestamp.slot - _gstate.last_producer_schedule_update.slot > 120 ) {
-         update_elected_producers( timestamp );
+         update_elected_producers( timestamp, bheader.previous );
```
Comment (jalbiero) [inline contracts/eosio.system/src/producer_pay.cpp]: @porkchop I don't know if this is correct (use _block_header::previous_ as the entry hash in our "random" system)

Comment (porkchop) [reply contracts/eosio.system/src/producer_pay.cpp]: Right now I think this block hash (the pointer to the previous block) is as good as any. Let's go with this at least for the time being

File: contracts/eosio.system/src/voting.cpp [hunk 2]
```diff
@@ -74,23 +92,55 @@ namespace eosiosystem {
...
       top_producers.reserve(21);
 
-      for ( auto it = idx.cbegin(); it != idx.cend() && top_producers.size() < 21 && 0 < it->total_votes && it->active(); ++it ) {
-         top_producers.emplace_back( std::pair<eosio::producer_key,uint16_t>({{it->owner, it->producer_key}, it->location}) );
+      if (standby_index < num_standbys) {
+         prod_vec_t standbys;
+         standbys.reserve(num_standbys);
+         select_producers_into(21, num_standbys, standbys);
+
+         if (standbys.size() == num_standbys)
```
Comment (porkchop) [inline contracts/eosio.system/src/voting.cpp]: if (standbys.size() > standby_index)

File: contracts/eosio.system/src/voting.cpp [hunk 3]
```diff
@@ -74,23 +92,55 @@ namespace eosiosystem {
...
+            top_producers.emplace_back(standbys[standby_index]);
       }
 
-      if ( top_producers.size() == 0 || top_producers.size() < _gstate.last_producer_schedule_size ) {
+      select_producers_into(0, 20, top_producers);
```
Comment (porkchop) [inline contracts/eosio.system/src/voting.cpp]: select_producers_into(0, 20 + top_producers.size(), top_producers)

Comment (jalbiero) [reply contracts/eosio.system/src/voting.cpp]: Correction: 
`select_producers_into(0, 21 - top_producers.size(), top_producers)`

File: contracts/eosio.system/src/producer_pay.cpp [hunk 4]
```diff
@@ -14,7 +14,10 @@ namespace eosiosystem {
 
       block_timestamp timestamp;
       name producer;
-      _ds >> timestamp >> producer;
+      uint16_t _;
+      checksum256 previous; 
+
+      _ds >> timestamp >> producer >> _ >> previous;
```
[the remark this answers is not in the document]
Comment (jalbiero) [reply contracts/eosio.system/src/producer_pay.cpp]: Yes, you're right, I've just followed the de-serialization flow :-P it makes no sense to perform an early deserialization of "previous" data.

File: contracts/eosio.system/src/voting.cpp [hunk 5]
```diff
@@ -12,6 +12,24 @@
...
+
+      uint64_t int_value = 0;
+      for (int i = 0; i < 8; i++) {
+         int_value <<= 8;
+         int_value |= static_cast<uint64_t>(byte_array[i]);
+      }
+      return int_value;
+   }
+
+   inline eosio::checksum256 next_random_value(const eosio::checksum256& value) {
```
[the remark this answers is not in the document]
Comment (jalbiero) [reply contracts/eosio.system/src/voting.cpp]: You're right again, that function belongs to the cold war era, it's not necessary anymore.

**Comment by jalbiero** [comment]
PR closed do to change in destination branch (_standby-producers_ instead of _develop_). See this new PR: https://github.com/worldwide-asset-exchange/wax-system-contracts/pull/34

**Pull request closed without merging**